sonar-core/src/main/java/org/sonar/core/issue/tracking/Tracker.java

```diff
@@ -0,0 +1,299 @@
+    SearchKey create(Trackable trackable);
+  }
+
+  private static class LineAndLineHashKey implements SearchKey {
+    private final RuleKey ruleKey;
+    private final String lineHash;
+    private final Integer line;
+
+    LineAndLineHashKey(Trackable trackable) {
+      this.ruleKey = trackable.getRuleKey();
+      this.line = trackable.getLine();
+      this.lineHash = StringUtils.defaultString(trackable.getLineHash(), "");
+    }
+
+    @Override
+    public boolean equals(@Nonnull Object o) {
+      if (this == o) {
+        return true;
+      }
+      LineAndLineHashKey that = (LineAndLineHashKey) o;
+      // start with most discriminant field
+      if (!Objects.equals(line, that.line)) {
+        return false;
+      }
+      if (!lineHash.equals(that.lineHash)) {
+        return false;
+      }
+      return ruleKey.equals(that.ruleKey);
+    }
+
+    @Override
+    public int hashCode() {
+      int result = ruleKey.hashCode();
+      result = 31 * result + lineHash.hashCode();
+      result = 31 * result + (line != null ? line.hashCode() : 0);
+      return result;
+    }
+  }
+
+  private enum LineAndLineHashKeyFactory implements SearchKeyFactory {
+    INSTANCE;
+    @Override
+    public SearchKey create(Trackable t) {
+      return new LineAndLineHashKey(t);
+    }
+  }
+
+  private static class LineHashAndMessageKey implements SearchKey {
+    private final RuleKey ruleKey;
+    private final String message, lineHash;
+
+    LineHashAndMessageKey(Trackable trackable) {
+      this.ruleKey = trackable.getRuleKey();
+      this.message = trackable.getMessage();
+      this.lineHash = StringUtils.defaultString(trackable.getLineHash(), "");
+    }
+
+    @Override
+    public boolean equals(@Nonnull Object o) {
+      if (this == o) {
+        return true;
+      }
+      LineHashAndMessageKey that = (LineHashAndMessageKey) o;
+      // start with most discriminant field
+      if (!lineHash.equals(that.lineHash)) {
+        return false;
+      }
+      if (!message.equals(that.message)) {
+        return false;
+      }
+      return ruleKey.equals(that.ruleKey);
+    }
+
+    @Override
+    public int hashCode() {
+      int result = ruleKey.hashCode();
+      result = 31 * result + message.hashCode();
+      result = 31 * result + lineHash.hashCode();
+      return result;
+    }
+  }
+
+  private enum LineHashAndMessageKeyFactory implements SearchKeyFactory {
+    INSTANCE;
+    @Override
+    public SearchKey create(Trackable t) {
+      return new LineHashAndMessageKey(t);
+    }
+  }
+
+  private static class LineAndMessageKey implements SearchKey {
+    private final RuleKey ruleKey;
+    private final String message;
+    private final Integer line;
+
+    LineAndMessageKey(Trackable trackable) {
+      this.ruleKey = trackable.getRuleKey();
+      this.message = trackable.getMessage();
+      this.line = trackable.getLine();
+    }
+
+    @Override
+    public boolean equals(@Nonnull Object o) {
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Add a type test to this method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2097)